pkg/controllermanager/controller/controllerregistration/controllerregistration_seed_control.go

```diff
@@ -269,46 +269,67 @@ func computeKindTypesForShoots(
+		alwaysPolicyControllerRegistrationNames []string
 	)
 
 	for _, controllerRegistration := range controllerRegistrationList {
 		controllerRegistrationNameToObject[controllerRegistration.Name] = controllerRegistration
+
 		for _, resource := range controllerRegistration.Spec.Resources {
-			kindTypeToControllerRegistrationName[fmt.Sprintf("%s/%s", resource.Kind, resource.Type)] = controllerRegistration.Name
+			id := fmt.Sprintf("%s/%s", resource.Kind, resource.Type)
+			kindTypeToControllerRegistrationNames[id] = append(kindTypeToControllerRegistrationNames[id], controllerRegistration.Name)
+		}
+
+		if controllerRegistration.Spec.Deployment != nil && controllerRegistration.Spec.Deployment.Policy != nil && *controllerRegistration.Spec.Deployment.Policy == gardencorev1beta1.ControllerDeploymentPolicyAlways {
+			alwaysPolicyControllerRegistrationNames = append(alwaysPolicyControllerRegistrationNames, controllerRegistration.Name)
 		}
 	}
 
-	return controllerRegistrationNameToObject, kindTypeToControllerRegistrationName
+	return controllerRegistrationNameToObject, kindTypeToControllerRegistrationNames, alwaysPolicyControllerRegistrationNames
 }
 
 // computeWantedControllerRegistrationNames computes the list of names of ControllerRegistration objects that are desired
 // to be installed. The computation is performed based on a list of required kind/type combinations and the proper mapping
-// to existing ControllerRegistration objects.
+// to existing ControllerRegistration objects. Additionally, all names in the alwaysPolicyControllerRegistrationNames list
+// will be returned.
 func computeWantedControllerRegistrationNames(
 	wantedKindTypeCombinations sets.String,
-	kindTypeToControllerRegistrationName map[string]string,
+	kindTypeToControllerRegistrationNames map[string][]string,
+	alwaysPolicyControllerRegistrationNames []string,
+	controllerRegistrationNameToObject map[string]*gardencorev1beta1.ControllerRegistration,
+	seedLabels map[string]string,
 ) (sets.String, error) {
-	names := sets.NewString()
+	controllerRegistrationNames := sets.NewString(alwaysPolicyControllerRegistrationNames...)
 
 	for _, requiredExtension := range wantedKindTypeCombinations.UnsortedList() {
-		name, ok := kindTypeToControllerRegistrationName[requiredExtension]
+		names, ok := kindTypeToControllerRegistrationNames[requiredExtension]
 		if !ok {
 			return nil, fmt.Errorf("need to install an extension controller for %q but no appropriate ControllerRegistration found", requiredExtension)
 		}
 
-		names.Insert(name)
+		controllerRegistrationNames.Insert(names...)
+	}
+
+	// filter controller registrations with non-matching seed selector
+	names, err := controllerRegistrationNamesWithMatchingSeedLabelSelector(controllerRegistrationNames.UnsortedList(), controllerRegistrationNameToObject, seedLabels)
+	if err != nil {
+		return nil, err
 	}
+	controllerRegistrationNames = controllerRegistrationNames.Intersection(names)
```

Comment:
Why do you need an intersect? controllerRegistrationNames contains all required names. Based on that you determine the ControllerRegistrations that have a matching seed selector. Is that not what you need already?

Reply:
Nice catch, yes, indeed, you're right. It was a left-over from intermediate work. I've removed it now.